Sources/HTTP/PoCSocket/PoCSocketConnectionListener.swift

```diff
@@ -195,10 +195,23 @@ public class PoCSocketConnectionListener: ParserConnecting {
 
     /// Starts reading from the socket and feeding that data to the parser
     public func process() {
-        try! socket?.setBlocking(mode: true)
-
-        let tempReaderSource = DispatchSource.makeReadSource(fileDescriptor: socket?.socketfd ?? -1,
-                                                             queue: socketReaderQueue)
+        let tempReaderSource: DispatchSourceRead
+        //Make sure we have a socket here.  Don't use guard so that
+        //  we don't encourage strongSocket to be used in the
+        //  event handler, which could cause a leak
+        if let strongSocket = socket {
+            do {
+                try strongSocket.setBlocking(mode: true)
+                tempReaderSource = DispatchSource.makeReadSource(fileDescriptor: strongSocket.socketfd,
+                                                                     queue: socketReaderQueue)
+            } catch {
+                print("Socket cannot be set to Blocking in process(): \(error)")
+                return
+            }
+        } else {
+            print("Socket is nil in process()")
+            return
+        }
 
         tempReaderSource.setEventHandler { [weak self] in
             guard let strongSelf = self else {
```

Comment:
For example, this line doesn't make sense to me because I don't think it can ever fail. `self` is never nil.

Reply:
The `self` is taken from the closure context (its not passed in as a parameter). As its only weakly referenced its not sufficient to keep it alive, so its possible that `self` no-longer exists.

This is a reasonable design pattern - I'm sure its used elsewhere.

Reply:
Thanks.  Wouldn't it be safer to set the event handler to `nil` when `self` is destroyed.  It's a class and that's what `deinit{}` is for?

In fact, we already have a `cleanup()` function which does exactly that, and it's called 4 lines below this.

I find it extremely difficult to reason about all this code.

Reply:
I didn't follow all this through, but if this is tightly coupled and one can't exist w/o the other, it should not be weak in the first place, but `[unowned]`. Weak is way more expensive for no gain in such a case.
(but be careful w/ threading issues)

Reply:
Yeah, so it comes down to whether the `readerSource` can (should) exist without its associated `PoCSocketConnectionListener` and `socket`. It looks on the surface like it shouldn't and that the lifecycle of each is tightly coupled. 

My guess is that this code is being conservative and avoiding a cyclic reference should `cleanup()` fail to get called (which in itself should be a bug).

Reply:
Not only is it possible for `self` to go away while the `readerSource` still exists, but it happens fairly often under load.  See the stack trace of the crash in #69 for just one example.

Reply:
OK and that's by design is it?

Reply:
It's not ideal, but with this API design, I think it's unavoidable.